README.txt

PdfLicenseManager README file
-----------------------------

Copyright (C) 2006  Enrico Masala, and
Copyright (C) 2008  Creative Commons (work by Asheesh Laroia).

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

Author: Enrico Masala   < masala _at-symbol_ polito dot it >
   and  Asheesh Laroia  < asheesh@creativecommons.org >

Date: 2008-04-08
Version: 2.3

Please cite the original author if you extend or modify this program.


INTRODUCTION
------------

This program aims to be a simple tool to manage pdf licensing information.
In particular, it is focused on inserting and managing Creative Commons licensing information.
For more info on Creative Commons, visit http://www.creativecommons.org

Please note that this program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

In short: use it at your own risk.


REQUIREMENTS
------------

To use the program, you need the itext library (which is able to manage pdf
files). The itext library can be downloaded at:
http://www.lowagie.com/iText/
You just need to download the jar file:
http://prdownloads.sourceforge.net/itext/itext-1.4.8.jar

This program has been developed using the itext library v. 1.4.8.
Please use itext v. 1.4.8 or higher.
This program has been developed using Java v. 1.4.2
Please use a Java version equal or higher than 1.4.2 .


LINUX platforms

First, download the itext-1.4.8.jar file into the current dir.

You can then run the program typing from this directory:

java -classpath itext-1.4.8.jar:. pdflicense.gui.MainWin

If the program is run without any parameter, it will show the graphical user
interface (GUI).
The program is still able to run in text mode.
Run it with one parameter (e.g. help) to get the help for textual mode.
Textual mode is useful to automate the process using, e.g., scripts.

In case you need to compile it, type:
javac -classpath itext-1.4.8.jar pdflicense/*.java  pdflicense/gui/*.java


OTHER platforms

This program has not been tested on other platforms.
However, if you have a Java 1.4.2 compliant environment (and you downloaded
the itext library), it should work, since it does not use any external or
particular features other than the itext library.

Note that, if you run java from the text interface of the command interpreter
in Windows, you need to use the ; instead of : in the -classpath argument.


EXAMPLES
--------

Show the XMP licensing info associated with a file.pdf:
Simply run the program without any parameter (to show the GUI) and open the
file.pdf using the File/Open menu

Insert the Creative Commons Attribution license into filein.pdf, creating fileout.pdf:
Once filein.pdf has been opened using the File/Open menu, select the license
and the version you want to insert, press the Update button, and then use the
File/Save as menu to save the file.

To run it in textual mode:

Show the XMP licensing info associated with a file.pdf:
java -classpath itext-1.4.8.jar:. pdflicense.ManagePdfLicense   show file.pdf

Insert the Creative Commons Attribution license into filein.pdf, creating fileout.pdf:
java -classpath itext-1.4.8.jar:. pdflicense.ManagePdfLicense   put filein.pdf fileout.pdf  by


HELP
----

Running the program with one option will show the following help screen:

PdfLicenseManager version 2.0, Copyright (C) 2006 Enrico Masala
PdfLicenseManager comes with ABSOLUTELY NO WARRANTY; for details see the enclosed LICENSE.txt file
This is free software, and you are welcome to redistribute it under certain conditions;
for details see the enclosed LICENSE.txt file

This software is licensed under GPL v.2.0
Please note that you are using it at your own risk

Usage: command  options

Commands:
help
                shows this help screen
showXMP    file.pdf
                shows the raw XMP info
show       file.pdf
                shows XMP licensing info
showToc    file.pdf
                shows pdf Toc (debug purposes only)
put        filein.pdf  fileout.pdf  <license>
                insert or update XMP licensing info, keep other info unchanged. <license> can be:
                by  by-nc  by-nd  by-sa  by-nc-nd  by-nc-sa
putforced  filein.pdf  fileout.pdf  <license>
                insert XMP licensing info removing any XMP info - use to correct corrupt XMP info
putXMP     filein.pdf  fileout.pdf  filein.xmp
                remove existing XMP and insert XMP description from file
                NB: for debug purposes only. Use put command to insert CC license



KNOWN BUGS

Since I am not an expert of XMP, the program may contain bugs or it may not work with certain files.
Always keep a copy of your original files.

Multiple XMP packets into a single pdf file are not managed by this program.



CHANGELOG.txt

CHANGELOG for the PdfLicenseManager program
-------------------------------------------
Author: Enrico Masala and Asheesh Laroia
See also SVN repository:
http://code.creativecommons.org/viewsvn/xmp/pdf_license_manager/trunk/pdflicense/

v.2.3   2008-04-06 (Asheesh Laroia)

Set license version to 3.0 by default
Refers to new Creative Commons namespace

v.2.2   Feb 7, 2006
Bug fix in updateLicense. It tried to append new nodes to the last
“rdf:Description” instead of "rdf:RDF". Thanks to Mr. dOGMaTiC and Andres Gonzalez.

v.2.1   Dec 4, 2006
Modified some info messages.

v.2.0   Jul 12, 2006
Added a graphical user interface. Command line interface is still present.

v.1.0   Jul 5, 2006
First release. Command line interface.



pdflicense/PdfLicenseManager.java
/*
 * This file is part of PdfLicenseManager.
 *
 * Copyright (C) 2006  Enrico Masala
 *
 * PdfLicenseManager is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * PdfLicenseManager is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Foobar; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 *
 * Author: Enrico Masala   < masala _at-symbol_ polito dot it >
 * Date: Jun 30, 2006
 *
 * Please cite the original author if you extend or modify this program
 */


package pdflicense;

import java.io.*;
import java.util.*;

import com.lowagie.text.*;
import com.lowagie.text.pdf.*;


public class PdfLicenseManager {

	public final static String version = "2.3";

	private String op = null;
	private String fNameIn = null;
	private String fNameOut = null;
	private String licenseShortName = null;
	private File fileIn = null;
	private File fileOut = null;
	private License currLic = null;
	private String xmpText = null;
	private XmpManager xm = null;
	private PdfReader reader = null;
	
	public String getFileNameIn() { return fNameIn; }
	public String getFileNameOut() { return fNameOut; }
	public File getFileIn() { return fileIn; }
	public File getFileOut() { return fileOut; }
	public String getLicenseShortName() { return licenseShortName; }
	public License getCurrLic() { return currLic; }
	public String getXmpText() { return xmpText; }
	
	public void setFileIn(File f) { fileIn = f; }
	public void setFileOut(File f) { fileOut = f; }
	
	public PdfLicenseManager() {
		xm = new XmpManager();
	}
	
	/**
	 * Print usage info
	 */
	private void usage() {
		System.out.print("\n");
		System.out.print("PdfLicenseManager version "+version+", Copyright (C) 2006 Enrico Masala\n");
		System.out.print("PdfLicenseManager comes with ABSOLUTELY NO WARRANTY; for details see the enclosed LICENSE.txt file\n");
		System.out.print("This is free software, and you are welcome to redistribute it under certain conditions;\n");
		System.out.print("for details see the enclosed LICENSE.txt file\n");
		System.out.print("\n");
		System.out.print("This software is licensed under GPL v.2.0\n");
		System.out.print("Please note that you are using it at your own risk\n");
		System.out.print("\n");
		System.out.print("Usage: command  options\n");
		System.out.print("\n");
		System.out.print("Commands:\n");
		System.out.print("help       \n                shows this help screen\n");
		System.out.print("showXMP    file.pdf\n                shows the raw XMP info\n");
		System.out.print("show       file.pdf\n                shows XMP licensing info\n");
		System.out.print("showToc    file.pdf\n                shows pdf Toc (debug purposes only)\n");
		System.out.print("put        filein.pdf  fileout.pdf  <license>\n                insert or update XMP licensing info, keep other info unchanged. <license> can be:\n");
		System.out.print("                by  by-nc  by-nd  by-sa  by-nc-nd  by-nc-sa\n");
		System.out.print("putforced  filein.pdf  fileout.pdf  <license>\n                insert XMP licensing info removing any XMP info - use to correct corrupt XMP info\n");
		System.out.print("putXMP     filein.pdf  fileout.pdf  filein.xmp\n                remove existing XMP and insert XMP description from file\n");
		System.out.print("                NB: It is safer to use put command to insert CC license\n");
		System.out.print("                    because it preserves the already present XMP info");
		System.out.print("\n\n");
	}

	/**
	 * Update the XMP info with the insertion of a CC licence
	 * 
	 * @param args input pdf file, output pdf file, xmp file
	 */
	public void mainTextual(String[] args) {


		if (args.length < 2) {
			usage();
			System.exit(1);
		} else {
			op = args[0];
			if (op.equals("show") || op.equals("showXMP") || op.equals("showToc")) {
				if (args.length != 2) {
					System.out.print("\n\nERROR: incorrect number of parameters");
					usage(); System.exit(1);
				}
				fNameIn = args[1];
			} else if (op.equals("put") || op.equals("putXMP") || op.equals("putforce") ) {
				if (args.length != 4) {
					System.out.print("\n\nERROR: incorrect number of parameters");
					usage(); System.exit(1);
				}
				fNameIn = args[1];
				fNameOut = args[2];
				licenseShortName = args[3];
			} else {
				System.out.print("ERROR: operation not known ("+op+")\n\n");
				usage(); System.exit(1);
			}
		}

		try {
			PdfReader reader = new PdfReader(fNameIn);
			XmpManager xm = new XmpManager();
			byte [] xmpdata = reader.getMetadata();

			if (op.equals("showXMP") || op.equals("show") ) {
				if (xmpdata==null) {
					System.out.print("No XMP Licensing info is present\n");
				} else {
					if (op.equals("showXMP")) {
						String s = new String(xmpdata);
						System.out.print(s);
					} else if (op.equals("show")) {
						String s = new String(xmpdata);
						xm.parseXmp(s);
						//System.out.println(xm.getXmpString());
						License currLic = xm.getLicenseInfo();
						System.out.print("***************************\n");
						System.out.print("XMP Licensing info:\n");
						System.out.print("***************************\n");
						System.out.print(currLic.toString());
						System.out.print("\n");
					}
				}
			} else if (op.equals("showToc")) {
				PdfDictionary cat = reader.getCatalog();
				Set k = cat.getKeys();	
				Object []arr = k.toArray();
				for (int i=0; i<arr.length; i++) {
					System.out.println(arr[i]);
				}
				//PdfObject p = cat.get(PdfName.METADATA);
				//System.out.println(p.getName());
				//System.out.println("/Metadata type="+p.type());
				
				//PRIndirectReference iref = (PRIndirectReference)p;
                		//findAllObjects(reader, PdfReader.getPdfObject(obj), hits);


				/*
				HashMap map = reader.getInfo();
				Set keys = map.keySet();	
				Object []arr = keys.toArray();
				for (int i=0; i<arr.length; i++) {
					System.out.println(arr[i]);
				}
				*/
			} else if (op.equals("put")) {
				CCLicense lic = new CCLicense(licenseShortName);
				if (xmpdata==null) {
					xm.createLicense(lic);
				} else {
					String s = new String(xmpdata);
					xm.parseXmp(s);
					xm.updateLicense(lic);
				}
				String xmpString = xm.getXmpString();
				FileOutputStream fos = new FileOutputStream(fNameOut);
				writeXmp(reader, fos, xmpString.getBytes());
			} else if (op.equals("putforce")) {
				CCLicense lic = new CCLicense(licenseShortName);
				xm.createLicense(lic);
				String xmpString = xm.getXmpString();
				FileOutputStream fos = new FileOutputStream(fNameOut);				
				writeXmp(reader, fos, xmpString.getBytes());
			} else if (op.equals("putXMP")) {
				File f = new File(licenseShortName);
				int length = (int)f.length();
				FileInputStream fin = new FileInputStream (f);
				BufferedInputStream data =  new BufferedInputStream(fin);
				byte[] xmp = new byte[length]; 
				data.read(xmp, 0, length);
				fin.close();		

				FileOutputStream fos = new FileOutputStream(fNameOut);
				writeXmp(reader, fos, xmp);
			}
		} catch (DocumentException de) {
			System.err.println(de.getMessage());
		} catch (IOException ioe) {
			System.err.println(ioe.getMessage());
		} catch (LicenseException e) {
			System.err.println(e.getMessage());
		}

	}

	
	private void writeXmp(PdfReader reader, FileOutputStream fos, byte[] xmpBytes) throws DocumentException, IOException {
		PdfStamper stamp = new PdfStamper(reader, fos);

		stamp.setXmpMetadata(xmpBytes);
		stamp.close();
	}

	public void readFileInfo() throws IOException, LicenseException {
		reader = new PdfReader(new FileInputStream(fileIn));
		byte [] xmpdata = reader.getMetadata();
		// reset current license info
		currLic = null;

		if (xmpdata != null) {
			xmpText = new String(xmpdata);
			xm.parseXmp(xmpText);
			//System.out.println(xm.getXmpString());
			currLic = xm.getLicenseInfo();
//			System.out.print("***************************\n");
//			System.out.print("XMP Licensing info:\n");
//			System.out.print("***************************\n");
//			System.out.print(currLic.toString());
//			System.out.print("\n");
		} else {
			xmpText = null;
		}
	}

	public void updateLicense(CCLicense lic) throws IOException, LicenseException {
		//System.out.print("updateLicense called\n");
		if (fileIn != null) {
			if (xmpText != null) {
				xm.updateLicense(lic);
			} else {
				xm.createLicense(lic);
			}
			xmpText = xm.getXmpString();
			xm.parseXmp(xmpText);
			currLic = xm.getLicenseInfo();
			//writeXmp(reader, fNameOut, xmpText.getBytes());
		}
	}

	public void writeFile(File f) throws DocumentException, IOException {
		FileOutputStream fos = new FileOutputStream(f);
		writeXmp(reader, fos, xmpText.getBytes());
	}
}
